android/src/main/kotlin/de/lschmierer/android_play_install_referrer/AndroidPlayInstallReferrerPlugin.kt

```diff
@@ -49,55 +33,64 @@ class AndroidPlayInstallReferrerPlugin : FlutterPlugin, MethodCallHandler {
 
     override fun onMethodCall(@NonNull call: MethodCall, @NonNull result: Result) {
         if (call.method == "getInstallReferrer") {
-            getInstallReferrer(call, result)
+            getInstallReferrer(result)
         } else {
             result.notImplemented()
         }
+    }
 
+    private fun getInstallReferrer(@NonNull result: Result) {
+        currentResult = result;
+        referrerClient = InstallReferrerClient.newBuilder(context).build()
+        referrerClient!!.startConnection(this)
     }
```

Comment:
What happens if user calls getInstallReferrer from Flutter again while the first request is still in-progress? Haven't tried, but the second request might overwrite referrerClient and currentResult, thus the first request never returns because the the first currentResult is never resolved?

I see two possible fixes:
a) Throwing an error if getInstallReferrer is called while in-progress (using an AtomicBoolean or Lock.tryLock()).
b) On second call dont overwrite referrerClient if referrerClient != null and collect all Results in a List and later resolve (call success or error) all Results. I think we still need some synchronization (e.g. new Result is pushed after currentResult.success/error is called but before referrerClient is set to null)

Aproach b) is probably considered cleaner API-wise. I can indeed imagine a developer trying to access the referrrer from two different locations in his Flutter codebase (although it might speak for a bad architecture on this develoeprs side).

Maybe you have some better/easier ideas?

Reply:
If my theory in https://github.com/lschmierer/android_play_install_referrer/pull/16#issuecomment-994756697 is correct, all this complicated logic is actually not needed.
Following https://github.com/lschmierer/android_play_install_referrer/pull/16#issuecomment-994756697 concurrency would not be an issue for us as we just forward it to the installreferrer API (just make to requests).

Reply:
From concurrency point of view you are right. So in general your approach was right. As for me the problem was in controlling listener instance and reference to it.